## 延伸题：改变交叉定标所用的模型，提升定标性能。

友情提醒：

可以探索的模型如对数函数、指数、函数、三次四次多项式、（全连接）神经网络等。这里附上一些你可能感兴趣的代码：

paras=np.polyfit(x,y,2) 该函数为多项式拟合；返回为降幂次系数的列表
如使用三次拟合，则改为 paras=np.polyfit(x,y,3)
将定标函数命名为 aimcalibrate。aimcalibrate=calibrate(waitArr,paras)，waitArr为需要定标图像，paras=np.polyfit(x,y,3) 为拟合后的各项系数。


【提升定标性能】的具体评价标准，指的是 R^2 比教案中更大、趋势与 GDP一致

定标技术熟练的同学，可以进一步尝试面到面的校正方式（而不是教案中点到点的那种）

#### 思路步骤

根据《灯光定标教案》进行修改，链接：https://www.heywhale.com/mw-org/meteoda/project/62f4aa3b337558d5db3e28a1

主要尝试修改定义函数：通过二次多项式拟合参考图像refArr与待定标图像waitArr之间的关系一段代码中
```python
paras=np.polyfit(x,y,3)
```
将np.polyfit(x,y,z)中的z设置为3就是三次拟合



### 预处理：准备数据

In [1]:
# 安装完成后，欢迎点右上角Kernel-保存当前镜像，以便以后直接使用该自定义镜像
!pip install retry pySmartDL -i https://pypi.douban.com/simple/ #从指定镜像下载安装工具包，镜像URL可自行修改

Looking in indexes: https://pypi.douban.com/simple/


In [2]:
# 导入需要的拓展包
from IPython.core.display import display, HTML
import tarfile  
import os
import gzip  
import shutil
import glob
from tqdm import tqdm
import gc
import gdal
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
%matplotlib inline 
import pandas as pd
from retry import retry
import pandas as pd
from PIL import Image
import imageio
import IPython
from pySmartDL import SmartDL

### 解压数据

In [3]:
# 定义函数：解压文件
@retry()
def gunzip_shutil(source_filepath, block_size=65536):
    # 解压gz文件，提取tif图像
    dest_filepath = source_filepath[:-3]
    with gzip.open(source_filepath, 'rb') as s_file, \
            open(dest_filepath, 'wb') as d_file:
        shutil.copyfileobj(s_file, d_file, block_size)

In [4]:
# 本cell由于数据量较大所以运行耗时较长
# 指定压缩文件gz的目录
inpDir = "/home/mw/input/NTL1992to20131028/stable_lights/"
# 获取所有压缩文件gz的文件名
gzFiles = glob.glob(inpDir + '*.gz')
gzFiles.sort()
# 依次遍历原始压缩文件gz
for gz in tqdm(gzFiles):
    # 从压缩文件gz中提取出stable_lights.avg_vis文件，该文件是tif格式的图像
    gunzip_shutil(gz)

print('全部解压成功！')

100%|██████████| 34/34 [02:34<00:00,  4.53s/it]

全部解压成功！


### 裁剪数据

In [5]:
# 定义函数：读取tif图像成numpy array
@retry()
def readTiff(filename:str):
    ds = gdal.Open(filename)
    image = ds.GetRasterBand(1).ReadAsArray()
    return image,ds.GetProjection(),ds.GetGeoTransform()

@retry()
# 定义函数：根据坐标裁剪子区域
def clipImg(image,proj,geotrans,upperLeftCoords,downRightCoords):
    # 根据坐标提取子矩阵
    r_upperLeft,c_upperLeft = upperLeftCoords
    r_downRight,c_downRight = downRightCoords
    subImage = image[r_upperLeft:r_downRight,c_upperLeft:c_downRight]
    # 设置投影信息
    subProj = proj
    # 设置坐标
    subGeotrans = list(geotrans)
    #图像左上角的坐标为r_upperLeft，c_upperLeft。则该点的经纬度为：
    xGeo = geotrans[0] + c_upperLeft * geotrans[1] + r_upperLeft * geotrans[2]
    yGeo = geotrans[3] + c_upperLeft * geotrans[4] + r_upperLeft * geotrans[5]
    # 左上角经度
    subGeotrans[0] = xGeo
    # 左上角纬度
    subGeotrans[3] = yGeo
    # 垂直空间分辨率(经度间隔)
    subGeotrans[1] = subGeotrans[1]
    # 水平空间分辨率(纬度间隔)
    subGeotrans[5] = subGeotrans[5]
    subGeotrans = tuple(subGeotrans)
    # 
    return subImage,subProj,subGeotrans

@retry()
# 定义函数：保存到磁盘
def outTiff(subArr,subProj,subGeotrans,outputFile):
    # 创建driver
    driver = gdal.GetDriverByName("GTiff")
    # 创建数据集
    outdata = driver.Create(outputFile, subArr.shape[1], subArr.shape[0], 1, gdal.GDT_Float32)
    # 把数组放进数据集
    outdata.GetRasterBand(1).WriteArray(subArr)
    # 设置投影信息和坐标
    outdata.SetProjection(subProj) 
    outdata.SetGeoTransform(subGeotrans)
    # 
    outdata.FlushCache()
    outdata = None
    driver=None
    subArr=None
    print('outTiff: {0}'.format(outputFile))

In [6]:
# 找到所有文件名
globalImgs = glob.glob('/home/mw/input/NTL1992to20131028/stable_lights/*tif')
globalImgs.sort()
# 定义输出路径
# 需提前挂载work目录
outDir = '/home/mw/work/2. clipped dataset/'
if not os.path.isdir(outDir):
    os.mkdir(outDir)
# 遍历所有文件
for fn in tqdm(globalImgs):
    # 读取数据
    image,proj,trans = readTiff(fn)
    # 裁剪图像
    subImage,subProj,subGeotrans = clipImg(image,proj,trans,(2433, 34070),(6971,38082)) #定义选取区间坐标
    del image
    _ = gc.collect()
    # 输出到磁盘
    bn = os.path.basename(fn)
    outFn = outDir + bn
    outTiff(subImage,subProj,subGeotrans,outFn)
    del subImage
    _ = gc.collect()
    # 为了节省空间，删除全球图像
    os.remove(fn)

  3%|▎         | 1/34 [00:14<08:14, 14.99s/it]

outTiff: /home/mw/work/2. clipped dataset/F101992.v4b_web.stable_lights.avg_vis.tif


  6%|▌         | 2/34 [00:23<06:55, 13.00s/it]

outTiff: /home/mw/work/2. clipped dataset/F101993.v4b_web.stable_lights.avg_vis.tif


  9%|▉         | 3/34 [00:29<05:43, 11.09s/it]

outTiff: /home/mw/work/2. clipped dataset/F101994.v4b_web.stable_lights.avg_vis.tif


 12%|█▏        | 4/34 [00:36<04:51,  9.72s/it]

outTiff: /home/mw/work/2. clipped dataset/F121994.v4b_web.stable_lights.avg_vis.tif


 15%|█▍        | 5/34 [00:43<04:20,  8.99s/it]

outTiff: /home/mw/work/2. clipped dataset/F121995.v4b_web.stable_lights.avg_vis.tif


 18%|█▊        | 6/34 [00:49<03:42,  7.96s/it]

outTiff: /home/mw/work/2. clipped dataset/F121996.v4b_web.stable_lights.avg_vis.tif


 21%|██        | 7/34 [00:55<03:19,  7.39s/it]

outTiff: /home/mw/work/2. clipped dataset/F121997.v4b_web.stable_lights.avg_vis.tif


 24%|██▎       | 8/34 [01:01<03:02,  7.03s/it]

outTiff: /home/mw/work/2. clipped dataset/F121998.v4b_web.stable_lights.avg_vis.tif


 26%|██▋       | 9/34 [01:08<02:51,  6.84s/it]

outTiff: /home/mw/work/2. clipped dataset/F121999.v4b_web.stable_lights.avg_vis.tif


 29%|██▉       | 10/34 [01:14<02:38,  6.62s/it]

outTiff: /home/mw/work/2. clipped dataset/F141997.v4b_web.stable_lights.avg_vis.tif


 32%|███▏      | 11/34 [01:20<02:32,  6.62s/it]

outTiff: /home/mw/work/2. clipped dataset/F141998.v4b_web.stable_lights.avg_vis.tif


 35%|███▌      | 12/34 [01:27<02:27,  6.69s/it]

outTiff: /home/mw/work/2. clipped dataset/F141999.v4b_web.stable_lights.avg_vis.tif


 38%|███▊      | 13/34 [01:33<02:13,  6.34s/it]

outTiff: /home/mw/work/2. clipped dataset/F142000.v4b_web.stable_lights.avg_vis.tif


 41%|████      | 14/34 [01:39<02:04,  6.21s/it]

outTiff: /home/mw/work/2. clipped dataset/F142001.v4b_web.stable_lights.avg_vis.tif


 44%|████▍     | 15/34 [01:45<02:01,  6.38s/it]

outTiff: /home/mw/work/2. clipped dataset/F142002.v4b_web.stable_lights.avg_vis.tif


 47%|████▋     | 16/34 [01:53<02:02,  6.81s/it]

outTiff: /home/mw/work/2. clipped dataset/F142003.v4b_web.stable_lights.avg_vis.tif


 50%|█████     | 17/34 [01:59<01:49,  6.43s/it]

outTiff: /home/mw/work/2. clipped dataset/F152000.v4b_web.stable_lights.avg_vis.tif


 53%|█████▎    | 18/34 [02:05<01:43,  6.44s/it]

outTiff: /home/mw/work/2. clipped dataset/F152001.v4b_web.stable_lights.avg_vis.tif


 56%|█████▌    | 19/34 [02:13<01:41,  6.74s/it]

outTiff: /home/mw/work/2. clipped dataset/F152002.v4b_web.stable_lights.avg_vis.tif


 59%|█████▉    | 20/34 [02:18<01:27,  6.24s/it]

outTiff: /home/mw/work/2. clipped dataset/F152003.v4b_web.stable_lights.avg_vis.tif


 62%|██████▏   | 21/34 [02:23<01:17,  5.98s/it]

outTiff: /home/mw/work/2. clipped dataset/F152004.v4b_web.stable_lights.avg_vis.tif


 65%|██████▍   | 22/34 [02:30<01:13,  6.15s/it]

outTiff: /home/mw/work/2. clipped dataset/F152005.v4b_web.stable_lights.avg_vis.tif


 68%|██████▊   | 23/34 [02:34<01:01,  5.55s/it]

outTiff: /home/mw/work/2. clipped dataset/F152006.v4b_web.stable_lights.avg_vis.tif


 71%|███████   | 24/34 [02:38<00:52,  5.20s/it]

outTiff: /home/mw/work/2. clipped dataset/F152007.v4b_web.stable_lights.avg_vis.tif


 74%|███████▎  | 25/34 [02:43<00:44,  4.99s/it]

outTiff: /home/mw/work/2. clipped dataset/F162004.v4b_web.stable_lights.avg_vis.tif


 76%|███████▋  | 26/34 [02:48<00:41,  5.25s/it]

outTiff: /home/mw/work/2. clipped dataset/F162005.v4b_web.stable_lights.avg_vis.tif


 79%|███████▉  | 27/34 [02:52<00:33,  4.79s/it]

outTiff: /home/mw/work/2. clipped dataset/F162006.v4b_web.stable_lights.avg_vis.tif


 82%|████████▏ | 28/34 [02:55<00:25,  4.32s/it]

outTiff: /home/mw/work/2. clipped dataset/F162007.v4b_web.stable_lights.avg_vis.tif


 85%|████████▌ | 29/34 [02:58<00:19,  3.85s/it]

outTiff: /home/mw/work/2. clipped dataset/F162008.v4b_web.stable_lights.avg_vis.tif


 88%|████████▊ | 30/34 [03:01<00:14,  3.66s/it]

outTiff: /home/mw/work/2. clipped dataset/F162009.v4b_web.stable_lights.avg_vis.tif


 91%|█████████ | 31/34 [03:04<00:09,  3.24s/it]

outTiff: /home/mw/work/2. clipped dataset/F182010.v4d_web.stable_lights.avg_vis.tif


 94%|█████████▍| 32/34 [03:09<00:07,  3.78s/it]

outTiff: /home/mw/work/2. clipped dataset/F182011.v4c_web.stable_lights.avg_vis.tif


 97%|█████████▋| 33/34 [03:11<00:03,  3.31s/it]

outTiff: /home/mw/work/2. clipped dataset/F182012.v4c_web.stable_lights.avg_vis.tif


100%|██████████| 34/34 [03:13<00:00,  5.68s/it]

outTiff: /home/mw/work/2. clipped dataset/F182013.v4c_web.stable_lights.avg_vis.tif


### 矫正F14

In [7]:
# 定义函数：将多个numpy array合并成一个长array
@retry()
def concat(listArr:"a list of 2d array",flatten:bool=True):
    arrConcat = np.concatenate(listArr,axis=0)
    if flatten:
        return arrConcat.flatten()
    return arrConcat

In [8]:
# 图像路径
fnF121997 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F121997*tif')[0]
fnF121998 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F121998*tif')[0]
fnF121999 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F121999*tif')[0]
fnF141997 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F141997*tif')[0]
fnF141998 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F141998*tif')[0]
fnF141999 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F141999*tif')[0]
# 读取F12的图像
arrF121997 = readTiff(fnF121997)[0]
print('{0} has been read.'.format('arrF121997'))
arrF121998 = readTiff(fnF121998)[0]
print('{0} has been read.'.format('arrF121998'))
arrF121999 = readTiff(fnF121999)[0]
print('{0} has been read.'.format('arrF121999'))
# 把arrF121997、arrF121998、arrF121999合并成一维数组
arrF12 = concat([arrF121997,arrF121998,arrF121999])
print('arrF121997,arrF121998,arrF121999 has been concated')
del arrF121997,arrF121998,arrF121999
_ = gc.collect()
# 把arrF141997、arrF141998、arrF141999合并成一维数组
arrF141997 = readTiff(fnF141997)[0]
print('{0} has been read.'.format('arrF141997'))
arrF141998 = readTiff(fnF141998)[0]
print('{0} has been read.'.format('arrF141998'))
arrF141999 = readTiff(fnF141999)[0]
print('{0} has been read.'.format('arrF141999'))
# 把arrF141997、arrF141998、arrF141999合并成一维数组
arrF14 = concat([arrF141997,arrF141998,arrF141999])
print('arrF141997,arrF141998,arrF141999 has been concated')
del arrF141997,arrF141998,arrF141999
_ = gc.collect()

arrF121997 has been read.
arrF121998 has been read.
arrF121999 has been read.
arrF121997,arrF121998,arrF121999 has been concated
arrF141997 has been read.
arrF141998 has been read.
arrF141999 has been read.
arrF141997,arrF141998,arrF141999 has been concated


In [9]:
# 定义函数：计算模型的决定系数R2
@retry()
def R2(paras,x,yTrue):
    #
    model = np.poly1d(paras)
    yPred = model(x)
    R2 = 1 - np.sum((yTrue - yPred)**2) / np.sum((yTrue - np.mean(yTrue))**2)
    return R2

# 定义函数：通过二次多项式拟合参考图像refArr与待定标图像waitArr之间的关系
#@retry()
def fitParameter(waitArr,refArr,xlabel='F14',ylabel='F12'):
    # 定义亮像元的掩膜mask，它指示参考图像和待定标图像中都亮的区域
    mask0 = waitArr>0
    mask1 = refArr>0
    mask = mask0 * mask1
    del mask0,mask1
    _ = gc.collect()
    # 应用mask
    x = waitArr[mask==1.]
    y = refArr[mask==1.]
    del mask
    _ = gc.collect()
    # 用二次多项式拟合参考图像和待定标图像之间的关系
    paras = np.polyfit(x, y, 3)
    # 绘图显示
    ## 点
    plt.figure(figsize=(5,5))
    from matplotlib.colors import LogNorm
    _ = plt.hist2d(x,y, bins=200, norm=LogNorm())
    plt.xlabel(xlabel)
    plt.ylabel(ylabel)
    ## 线 
    line  = []
    for i in range(64):
        line.append(np.poly1d(paras)(i))
    plt.plot(list(range(64)),line,'-')
    ## 参数
    plt.text(70, 10, 'w2={0}\n\nw1={1}\n\nw0={2}'.format(paras[0],paras[1],paras[2]))
    r_sq = R2(paras,x,y)
    plt.text(70, 4, 'R2={0}'.format(r_sq))
    del x,y
    _ = gc.collect()
    # 返回参数和R2
    return paras,r_sq

In [10]:
waitArr = arrF14
refArr = arrF12
paras,r_sq = fitParameter(waitArr,refArr)
print('parameters: {0}'.format(paras))
print('R2: {0}'.format(r_sq))
del waitArr,refArr,arrF14,arrF12
_ = gc.collect()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: RankWarning: Polyfit may be poorly conditioned
  This is separate from the ipykernel package so we can avoid doing imports until


parameters: [1.31270056e-04 7.53685874e-03 3.45877729e-01 3.00352788e+00]
R2: 0.7890328019857407


<Figure size 360x360 with 1 Axes>

In [11]:
# 定义函数：根据拟合的参数对待定标图像进行定标
def calibrate(waitArr,paras):
    model = np.poly1d(paras)
    waitArr[waitArr!=0.] = model(waitArr[waitArr!=0.])
    waitArr[waitArr<0.] = 0.
    waitArr[waitArr>63.] = 63.
    return waitArr

# 定义函数：输出图像
def outTif(arr,outFn,proj,trans):
    # 创建driver
    nrow = arr.shape[0]
    ncol = arr.shape[1]
    driver = gdal.GetDriverByName("GTiff")
    outdata = driver.Create(outFn,ncol,nrow,1, gdal.GDT_Float32)
    outband = outdata.GetRasterBand(1)
    # 写出数组
    outband.WriteArray(arr)
    # 添加地理参考
    im_proj = proj
    im_geotrans = trans
    outdata.SetGeoTransform(im_geotrans)
    outdata.SetProjection(im_proj)
    driver = None
    outdata = None
    outband = None


In [12]:
# 创建定标图像的输出路径
calibratedDir = '/home/mw/work/3. calibrated dataset/'
if not os.path.isdir(calibratedDir):
    os.mkdir(calibratedDir)

# 收集F14所有的图像
F14s = ['F141997','F141998','F141999','F142000','F142001','F142002','F142003']

# 遍历每一幅图像
for F14xxxx in F14s:
    # 读取图像，对其进行定标
    fnF14xxxx = glob.glob('/home/mw/work/2. clipped dataset/' + '*' + F14xxxx + '*tif')[0]
    arrF14xxxx,proj,trans = readTiff(fnF14xxxx)
    arrF14xxxx = calibrate(arrF14xxxx,paras)
    # 将定标的F14xxxx图像输出
    outFn = calibratedDir + os.path.basename(fnF14xxxx)
    outTif(arrF14xxxx,outFn,proj,trans)
    print('{0} has been output'.format(outFn))
    del arrF14xxxx
    _ = gc.collect()

/home/mw/work/3. calibrated dataset/F141997.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F141998.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F141999.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F142000.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F142001.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F142002.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F142003.v4b_web.stable_lights.avg_vis.tif has been output


In [13]:
# 找到所有F10卫星的图像
F10s = glob.glob('/home/mw/work/2. clipped dataset/' + 'F10*tif')
F10s.sort()
# 找到所有F12卫星的图像
F12s = glob.glob('/home/mw/work/2. clipped dataset/' + 'F12*tif')
F12s.sort()
# 依次拷贝到/3. calibrated dataset/下面
for f1nxxxx in tqdm(F10s + F12s):
    shutil.copy(f1nxxxx,  '/home/mw/work/3. calibrated dataset/'+os.path.basename(f1nxxxx))

100%|██████████| 9/9 [00:08<00:00,  1.07it/s]


In [14]:
# 定义函数：灯光亮度求和
def sumNTL(arr):
    sumNTL = np.sum(arr)
    return sumNTL

# 定义函数：创建dataframe以保存灯光总和值
def createDF():
    df = pd.DataFrame({"F10":[np.NAN for i in range(1992,2014)],
                       "F12":[np.NAN for i in range(1992,2014)],
                       "F14":[np.NAN for i in range(1992,2014)],
                       "F15":[np.NAN for i in range(1992,2014)],
                       "F16":[np.NAN for i in range(1992,2014)],
                       "F18":[np.NAN for i in range(1992,2014)],},index=range(1992,2014))
    return df

In [15]:
# 统计F10、F12、F14卫星的定标以前的数据（位于/2. clipped dataset/下）的灯光总和
dfSum_beforeCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/2. clipped dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/2. clipped dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/2. clipped dataset/F14*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_beforeCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [16]:
# 统计F10、F12、F14卫星的定标以后的数据（位于/3. calibrated dataset/下）的灯光总和
dfSum_afterCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F14*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_afterCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [17]:
dfSum_beforeCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !BEFORE! calibration')
dfSum_afterCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !AFTER! calibration')

<AxesSubplot:title={'center':'sum of light values !AFTER! calibration'}>

<Figure size 720x216 with 1 Axes>

<Figure size 720x216 with 1 Axes>

### 矫正F15

In [18]:
# 图像路径
fnF142003 = glob.glob('/home/mw/work/3. calibrated dataset/'+'*F142003*tif')[0]
fnF152003 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F152003*tif')[0]
# 读取F142003的图像
arrF142003 = readTiff(fnF142003)[0]
# 把F142003数组展成一维数组
arrF142003 = concat([arrF142003])
# 读取F152003的图像
arrF152003 = readTiff(fnF152003)[0]
# 把F152003数组展成一维数组
arrF152003 = concat([arrF152003])

In [19]:
waitArr = arrF152003
refArr = arrF142003
paras,r_sq = fitParameter(waitArr,refArr,xlabel='F152003',ylabel='F142003')
print('parameters: {0}'.format(paras))
print('R2: {0}'.format(r_sq))
del waitArr,refArr,arrF152003,arrF142003
_ = gc.collect()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: RankWarning: Polyfit may be poorly conditioned
  This is separate from the ipykernel package so we can avoid doing imports until


parameters: [1.12517807e-04 6.59801607e-03 3.41667932e-01 4.19049053e+00]
R2: 0.9208144545555115


<Figure size 360x360 with 1 Axes>

In [20]:
# 输出路径
calibratedDir = '/home/mw/work/3. calibrated dataset/'
if not os.path.isdir(calibratedDir):
    os.mkdir(calibratedDir)

# 收集F14所有的图像
F15s = ['F152003','F152004','F152005','F152006','F152007']

# 遍历每一幅图像
for F15xxxx in F15s:
    # 读取图像，对其进行定标
    fnF15xxxx = glob.glob('/home/mw/work/2. clipped dataset/' + '*' + F15xxxx + '*tif')[0]
    arrF15xxxx,proj,trans = readTiff(fnF15xxxx)
    arrF15xxxx = calibrate(arrF15xxxx,paras)
    # 将定标的F15xxxx图像输出
    outFn = calibratedDir + os.path.basename(fnF15xxxx)
    outTif(arrF15xxxx,outFn,proj,trans)
    print('{0} has been output'.format(outFn))
    del arrF15xxxx
    _ = gc.collect()

/home/mw/work/3. calibrated dataset/F152003.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F152004.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F152005.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F152006.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F152007.v4b_web.stable_lights.avg_vis.tif has been output


In [21]:
# 找到所有F15卫星的图像
F15s =  glob.glob('/home/mw/work/2. clipped dataset/' + 'F152000*tif')+ \
        glob.glob('/home/mw/work/2. clipped dataset/' + 'F152001*tif')+ \
        glob.glob('/home/mw/work/2. clipped dataset/' + 'F152002*tif')
F15s.sort()
# 依次拷贝到/3. calibrated dataset/下面
for f1nxxxx in tqdm(F15s):
    shutil.copy(f1nxxxx,  '/home/mw/work/3. calibrated dataset/'+os.path.basename(f1nxxxx))

100%|██████████| 3/3 [00:03<00:00,  1.00s/it]


In [22]:
# 统计F10、F12、F14、F15卫星的定标以前的数据（位于/2. clipped dataset/下）的灯光总和
dfSum_beforeCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/2. clipped dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/2. clipped dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/2. clipped dataset/F14*tif')
fnF15xxxx = glob.glob('/home/mw/work/2. clipped dataset/F15*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx+ fnF15xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_beforeCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [23]:
# 统计F10、F12、F14卫星的定标以后的数据（位于/3. calibrated dataset/下）的灯光总和
dfSum_afterCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F14*tif')
fnF15xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F15*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx + fnF15xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_afterCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [24]:
dfSum_beforeCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !BEFORE! calibration')
dfSum_afterCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !AFTER! calibration')

<AxesSubplot:title={'center':'sum of light values !AFTER! calibration'}>

<Figure size 720x216 with 1 Axes>

<Figure size 720x216 with 1 Axes>

### 矫正F16

In [25]:
# 图像路径
fnF152005 = glob.glob('/home/mw/work/3. calibrated dataset/'+'*F152005*tif')[0]
fnF162005 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F162005*tif')[0]
# 读取F152005的图像
arrF152005 = readTiff(fnF152005)[0]
# 把F152005数组展成一维数组
arrF152005 = concat([arrF152005])
# 读取F162005的图像
arrF162005 = readTiff(fnF162005)[0]
# 把F162005数组展成一维数组
arrF162005 = concat([arrF162005])

In [26]:
waitArr = arrF162005
refArr = arrF152005
paras,r_sq = fitParameter(waitArr,refArr,xlabel='F162005',ylabel='F152005')
print('parameters: {0}'.format(paras))
print('R2: {0}'.format(r_sq))
del waitArr,refArr,arrF162005,arrF152005
_ = gc.collect()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: RankWarning: Polyfit may be poorly conditioned
  This is separate from the ipykernel package so we can avoid doing imports until


parameters: [9.71613381e-05 5.84008982e-03 3.22406807e-01 4.49176068e+00]
R2: 0.9545505121350288


<Figure size 360x360 with 1 Axes>

In [27]:
calibratedDir = '/home/mw/work/3. calibrated dataset/'
if not os.path.isdir(calibratedDir):
    os.mkdir(calibratedDir)

# 收集F14所有的图像
F16s = ['F162004','F162005','F162006','F162007','F162008','F162009']

# 遍历每一幅图像
for F16xxxx in F16s:
    # 读取图像，对其进行定标
    fnF16xxxx = glob.glob('/home/mw/work/2. clipped dataset/' + '*' + F16xxxx + '*tif')[0]
    arrF16xxxx,proj,trans = readTiff(fnF16xxxx)
    arrF16xxxx = calibrate(arrF16xxxx,paras)
    # 将定标的F16xxxx图像输出
    outFn = calibratedDir + os.path.basename(fnF16xxxx)
    outTif(arrF16xxxx,outFn,proj,trans)
    print('{0} has been output'.format(outFn))
    del arrF16xxxx
    _ = gc.collect()

/home/mw/work/3. calibrated dataset/F162004.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F162005.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F162006.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F162007.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F162008.v4b_web.stable_lights.avg_vis.tif has been output
/home/mw/work/3. calibrated dataset/F162009.v4b_web.stable_lights.avg_vis.tif has been output


In [28]:
# 统计F10、F12、F14、F15、F16卫星的定标以前的数据（位于/2. clipped dataset/下）的灯光总和
dfSum_beforeCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/2. clipped dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/2. clipped dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/2. clipped dataset/F14*tif')
fnF15xxxx = glob.glob('/home/mw/work/2. clipped dataset/F15*tif')
fnF16xxxx = glob.glob('/home/mw/work/2. clipped dataset/F16*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx+ fnF15xxxx+ fnF16xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_beforeCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [29]:
# 统计F10、F12、F14、F15、F16卫星的定标以后的数据（位于/3. calibrated dataset/下）的灯光总和
dfSum_afterCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F14*tif')
fnF15xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F15*tif')
fnF16xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F16*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx + fnF15xxxx+ fnF16xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_afterCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [30]:
dfSum_beforeCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !BEFORE! calibration')
dfSum_afterCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !AFTER! calibration')

<AxesSubplot:title={'center':'sum of light values !AFTER! calibration'}>

<Figure size 720x216 with 1 Axes>

<Figure size 720x216 with 1 Axes>

#### 校正F182010

In [31]:
# 图像路径
fnF162009 = glob.glob('/home/mw/work/3. calibrated dataset/'+'*F162009*tif')[0]
fnF182010 = glob.glob('/home/mw/work/2. clipped dataset/'+'*F182010*tif')[0]
# 读取F162009的图像
arrF162009 = readTiff(fnF162009)[0]
# 把F162009数组展成一维数组
arrF162009 = concat([arrF162009])
# 读取F182010的图像
arrF182010 = readTiff(fnF182010)[0]
# 把F182010数组展成一维数组
arrF182010 = concat([arrF182010])

In [32]:
waitArr = arrF182010
refArr = arrF162009
paras,r_sq = fitParameter(waitArr,refArr,xlabel='F182010',ylabel='F162009')
print('parameters: {0}'.format(paras))
print('R2: {0}'.format(r_sq))
del waitArr,refArr,arrF182010,arrF162009
_ = gc.collect()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: RankWarning: Polyfit may be poorly conditioned
  This is separate from the ipykernel package so we can avoid doing imports until


parameters: [7.72280084e-05 4.48685232e-03 2.19147042e-01 3.69557055e+00]
R2: 0.9015113413333893


<Figure size 360x360 with 1 Axes>

In [33]:
# 输出路径
calibratedDir = '/home/mw/work/3. calibrated dataset/'
if not os.path.isdir(calibratedDir):
    os.mkdir(calibratedDir)

# 读取图像，对其进行定标
fnF182010 = glob.glob('/home/mw/work/2. clipped dataset/' + '*F182010*tif')[0]
arrF182010,proj,trans = readTiff(fnF182010)
arrF182010 = calibrate(arrF182010,paras)
# 将定标的F182010图像输出
outFn = calibratedDir + os.path.basename(fnF182010)
outTif(arrF182010,outFn,proj,trans)
print('{0} has been output'.format(outFn))
del arrF182010
_ = gc.collect()

/home/mw/work/3. calibrated dataset/F182010.v4d_web.stable_lights.avg_vis.tif has been output


In [34]:
# 找到所有F18卫星的图像
F18s =  glob.glob('/home/mw/work/2. clipped dataset/' + 'F182011*tif')+ \
        glob.glob('/home/mw/work/2. clipped dataset/' + 'F182012*tif')+ \
        glob.glob('/home/mw/work/2. clipped dataset/' + 'F182013*tif')
F18s.sort()
# 依次拷贝到/3. calibrated dataset/下面
for f1nxxxx in tqdm(F18s):
    shutil.copy(f1nxxxx,  '/home/mw/work/3. calibrated dataset/'+os.path.basename(f1nxxxx))

100%|██████████| 3/3 [00:03<00:00,  1.19s/it]


In [35]:
# 统计F10、F12、F14、F15、F18卫星的定标以前的数据（位于/2. clipped dataset/下）的灯光总和
dfSum_beforeCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/2. clipped dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/2. clipped dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/2. clipped dataset/F14*tif')
fnF15xxxx = glob.glob('/home/mw/work/2. clipped dataset/F15*tif')
fnF16xxxx = glob.glob('/home/mw/work/2. clipped dataset/F16*tif')
fnF18xxxx = glob.glob('/home/mw/work/2. clipped dataset/F18*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx+ fnF15xxxx+ fnF16xxxx+ fnF18xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_beforeCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [36]:
# 统计F10、F12、F14、F15、F16、F18卫星的定标以后的数据（位于/3. calibrated dataset/下）的灯光总和
dfSum_afterCalibration = createDF()
# 遍历每一张图，进行求和计算
fnF10xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F10*tif')
fnF12xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F12*tif')
fnF14xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F14*tif')
fnF15xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F15*tif')
fnF16xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F16*tif')
fnF18xxxx = glob.glob('/home/mw/work/3. calibrated dataset/F18*tif')
for fn in fnF10xxxx + fnF12xxxx + fnF14xxxx + fnF15xxxx+ fnF16xxxx+ fnF18xxxx:
    # 读取图像计算总和
    arr = readTiff(fn)[0]
    sumValue = sumNTL(arr)
    # 保存到dataframe的相应位置
    fnn = os.path.basename(fn)[0:3]
    xxxx = int(os.path.basename(fn)[3:7])
    dfSum_afterCalibration.loc[xxxx,fnn] = sumValue
    del arr
    gc.collect()

In [37]:
dfSum_beforeCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !BEFORE! calibration')
dfSum_afterCalibration.plot(kind='line',figsize=[10,3],title='sum of light values !AFTER! calibration')

<AxesSubplot:title={'center':'sum of light values !AFTER! calibration'}>

<Figure size 720x216 with 1 Axes>

<Figure size 720x216 with 1 Axes>

## 总结

原始课件中的R2数值：

F14:0.9085130915045738
F15:0.9481293857097626
F16:0.9603672511875629
F182010:0.9080354273319244


更改拟合次数为3后的R2数值：
F14:0.7890328019857407
F15:0.9208144545555115
F16:0.9545505121350288
F182010:0.9015113413333893

整体学习对于气象模型的了解还需要加深，课余时间再更多的了解一下气象模型后未来对于这个模型回头再尝试更换更多的模型和方法后再看看R2的数值变化